### PR TITLE
tests/riotboot_flashwrite: add python test script

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds a test script for `riotboot_flashwrite`. ~~It it still missing a lot of dependencies PR's from @kaspar030.~~

~~NOTE: the autthorship in the first 6 commits is only @kaspar030 it's a residue of `cherry-pick`.~~

### Testing procedure

Follow the README and test over ethos and wireless.

### Issues/PRs references

Depends on ~~https://github.com/RIOT-OS/RIOT/pull/11707~~,  ~~https://github.com/RIOT-OS/RIOT/pull/11690~~ and ~~https://github.com/RIOT-OS/RIOT/pull/11816~~.

### CODE REVIEW

**Comment by fjmolinas** [comment]
@kaspar030 there are no dependencies anymore!

File: tests/riotboot_flashwrite/Makefile [hunk 1]
```diff
@@ -1,6 +1,8 @@
-DEVELHELP ?= 0
+# If no BOARD is found in the environment, use this default:
```
Comment (kaspar030) [inline tests/riotboot_flashwrite/Makefile]: tests usually include ```../Makefile.tests_common```.
That sets BOARD and DEVELHELP to defaults, and also configures the application name to something that cannot clash. As is, this application would be called ```riotboot_flashwrite```, same as the module.

Comment (fjmolinas) [reply tests/riotboot_flashwrite/Makefile]: It is still included I just moved it to the bottom. The first commit only does that, to group things better (although that is just my personal taste I guess), but I can always drop that commit.

Comment (kaspar030) [reply tests/riotboot_flashwrite/Makefile]: Ahok. Then please move it back to the top, for consistency. Just put the variables that override it above.

**Comment by fjmolinas** [comment]
> Having the suit testing in the queue, which implicitly tests riotboot_flashwrite, does it make sense to have another test application?

IMO although suit implicitly tests `riotboot_flashwrite`, having an explicit test is better. Tests that cover specific or smaller building block are needed even if there is an integration test that covers the whole system. Following that argument why would I have a test for `sx1276` when `semtech_loramac` already tests that. If I'm testing suit for a platform if see the test fails I can go back on the building blocks, does `tests/riotboot` succeed? does `tests/riotboot_flashwrite` succeed? IMO being able to isolate things helps with testing and finding bugs.

Also, although suit attempts to be a standard, it might not be used, people might still implement there own OTA mechanism, because they want something smaller/simplet, or because they think having something privative is more secure (I'm not stating an opinion here), or whatever reason. I think saying a `suit` test is enough is presuming on what a user or developer might want to do with `riotboot_flashwrite` or `riotboot`.

I also think that running `tests/riotboot_flashwrite` should be part of the standard testing a user does when implementing riotboot for a board (if it supports it). `tests/riotboot` could succeed and `tests/riotboot_flashwrite` fail because of alignment issues between the slots/bootloader and the `flashpage_size`, etc.

And as a developer I find it useful do have an automatic script I can launch whenever I change `riotboot_flashwrite` (completely biased opinion :) ).

File: tests/riotboot_flashwrite/Makefile (hunk 1 above)
Comment (fjmolinas) [reply tests/riotboot_flashwrite/Makefile]: Done.

**Comment by kaspar030** [comment]
BTW, if you'd like to test just this test on murdock without rebuilding, you can add a "REMOVEME" commit that adds a filter to "./.murdock", somewhere at the top:

```
BOARDS=samr21-xpro
APPS=tests/riotboot_flashwrite
```

That should limit the build to that one application/board.

**Comment by kaspar030** [comment]
> And as a developer I find it useful do have an automatic script I can launch whenever I change `riotboot_flashwrite` (completely biased opinion :) ).

OK! (I just saw like 80% duplication in Makefile and test...)

**Comment by kaspar030** [comment]
> (I just saw like 80% duplication in Makefile and test...)

(and the flash/reflash tests over network block the board for >two minutes, compared to ~20sec for a regular test)

**Comment by fjmolinas** [comment]
> (and the flash/reflash tests over network block the board for >two minutes, compared to ~20sec for a regular test)

> OK! (I just saw like 80% duplication in Makefile and test...)

Yes I know... :( maybe the solution could be to modify the suit test.. so it proceeds with the update even if some part of suit verification fails? The test scipt would fail but only on the suit part and not the flashwrite part?

**Comment by fjmolinas** [comment]
> BTW, if you'd like to test just this test on murdock without rebuilding, you can add a "REMOVEME" commit that adds a filter to "./.murdock", somewhere at the top:
> 
> BOARDS=samr21-xpro
> APPS=tests/riotboot_flashwrite
> That should limit the build to that one application/board.

Cool! Thanks!

File: tests/riotboot_flashwrite/Makefile [hunk 2]
```diff
@@ -55,9 +73,19 @@ ifneq (,$(filter $(BOARD),$(LOW_MEMORY_BOARDS)))
...
+# So use that here in order to prevent the softdevice from being used.
+ifeq (nrf52dk,$(BOARD))
+  USEMODULE += nrfmin
+endif
+
+# The test needs the linked slot binaries without header in order to be able to
+# create final binaries with specific APP_VER values. The CI RasPi test workers
+# don't compile themselves, thus add the required files here so they will be
+# submitted along with the test jobs.
+TEST_EXTRA_FILES=$(SLOT_RIOT_ELFS)
```
Comment (aabadie) [inline tests/riotboot_flashwrite/Makefile]:
```suggestion
TEST_EXTRA_FILES = $(SLOT_RIOT_ELFS)
```

File: tests/riotboot_flashwrite/main.c [hunk 3]
```diff
@@ -36,8 +36,20 @@ int main(void)
     puts("riotboot_flashwrite test application");
 
     int current_slot = riotboot_slot_current();
-    printf("Current slot=%d\n", current_slot);
-    riotboot_slot_print_hdr(current_slot);
+    if (current_slot != -1) {
+        /* Sometimes, udhcp output messes up the following printfs.  That
```
Comment (aabadie) [inline tests/riotboot_flashwrite/main.c]:
```suggestion
        /* Sometimes, udhcp output messes up the following printfs. That
```

File: tests/riotboot_flashwrite/tests/01-run.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        "coap-client",
+        "-m",
+        "post",
+        "coap://{}/flashwrite".format(client_url),
+        "-f",
+        "tests_riotboot_flashwrite-slot{}.{}.riot.bin".format(slot, version),
+        "-b",
+        "64",
+    ]
+    subprocess.Popen(cmd, cwd=BINDIR)
```
Comment (aabadie) [inline tests/riotboot_flashwrite/tests/01-run.py]: Why not :
```suggestion
    subprocess.call(cmd, cwd=BINDIR)
```
?
with eventually the use of an assert.

File: tests/riotboot_flashwrite/tests/01-run.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        "riotboot",
+    ]
+    assert not subprocess.call(cmd)
+
+
+def testfunc(child):
+    """For one board test if specified application is updatable"""
+
+    # Initial Setup and wait for address configuration
+    child.expect_exact("main(): This is RIOT!")
```
Comment (aabadie) [inline tests/riotboot_flashwrite/tests/01-run.py]: This is not needed now with the `test_interactive_sync` module.

**Comment by fjmolinas** [comment]
Rebased but there are some issue with the script will address later.

**Comment by aabadie** [comment]
still useful but might need a refresh

**Comment by fjmolinas** [comment]
Staled, doesn't work anymore.

**Comment by aabadie** [comment]
> Staled, doesn't work anymore.

What is not working ?

**Comment by fjmolinas** [comment]
> > Staled, doesn't work anymore.
> 
> What is not working ?

See https://github.com/RIOT-OS/RIOT/pull/11808#issuecomment-573773054

**Comment by aabadie** [comment]
> See #11808 (comment)

You mean "there are some issue with the script will address later." This is a bit vague ;)

**Pull request closed without merging**